### PR TITLE
Add local PV negative scheduling tests to integration testing

### PR DESCRIPTION
**Pull request by sbezverk**
Closes: #56088

```release-note
Move local PV negative scheduling tests to integration
```

### CODE REVIEW

File: test/integration/scheduler/local-pv-neg-affinity_test.go [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+		}
+		fmt.Printf("%s\n", printIndentedJson(resultingPod))
+
+		if err := config.client.CoreV1().Pods(config.ns).Delete("local-pod", &metav1.DeleteOptions{}); err != nil {
+			t.Fatalf("Failed to create Pod %q: %v", pod.Name, err)
+		}
+	}
+}
+
+func setupNodes(t *testing.T, nsName string, numberOfNodes int) *testConfig {
```
Comment (msau42) [inline test/integration/scheduler/local-pv-neg-affinity_test.go]: Are you able to reuse the setup function in the volume binding test?

File: test/integration/scheduler/local-pv-neg-affinity_test.go [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+	for i := 0; i < len(nodeMarkers); i++ {
+		pod := makePod("local-pod", config.ns, []string{"local-pvc"})
+		nodeMarkers[i].(func(*v1.Pod, string))(pod, "node-2")
+		// Create Pod
+		if _, err := config.client.CoreV1().Pods(config.ns).Create(pod); err != nil {
+			t.Fatalf("Failed to create Pod %q: %v", pod.Name, err)
+		}
+
+		if err := waitForPodToSchedule(config.client, pod); err != nil {
```
Comment (msau42) [inline test/integration/scheduler/local-pv-neg-affinity_test.go]: Can you validate that the pod failed to schedule?

It would be good if we could validate the Pod events and look for the predicate filters.  You may need to add some logic in the setup to start the scheduler with an event recorder.

File: test/integration/scheduler/local-pv-neg-affinity_test.go [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+
+	// Create PVC
+	if _, err := config.client.CoreV1().PersistentVolumeClaims(config.ns).Create(pvc); err != nil {
+		t.Fatalf("Failed to create PersistentVolumeClaim %q: %v", pvc.Name, err)
+	}
+
+	nodeMarkers := []interface{}{
+		markNodeAffinity,
+		markNodeSelector,
+		markNodeName,
```
Comment (msau42) [inline test/integration/scheduler/local-pv-neg-affinity_test.go]: This test case won't work here because it actually completely bypasses the scheduler, and it will fail on the kubelet side.  So this last test case needs to remain in the e2e test for now.

Can you also delete the 1st two test cases from the e2e test suite?

File: test/integration/scheduler/local-pv-neg-affinity_test.go (hunk 1 above)
Comment (sbezverk) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: If you agree, I will modify setup functions to be able to start mulitple nodes. In this case in your tests I will call it with number of nodes == 1

File: test/integration/scheduler/local-pv-neg-affinity_test.go (hunk 3 above)
Comment (sbezverk) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: will do

File: test/integration/scheduler/local-pv-neg-affinity_test.go (hunk 2 above)
Comment (sbezverk) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: done

File: test/integration/scheduler/local-pv-neg-affinity_test.go (hunk 1 above)
Comment (sbezverk) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: done

Comment (sbezverk) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: When I use new seupNodes in yout volume binding test, both tests fail, when I use in my PR new setupNodes and in yours setup, both tests pass. Mystery...

File: test/integration/scheduler/local-pv-neg-affinity_test.go [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+	// Create PVC
+	if _, err := config.client.CoreV1().PersistentVolumeClaims(config.ns).Create(pvc); err != nil {
+		t.Fatalf("Failed to create PersistentVolumeClaim %q: %v", pvc.Name, err)
+	}
+
+	nodeMarkers := []interface{}{
+		markNodeAffinity,
+		markNodeSelector,
+	}
+	podName := ""
```
Comment (msau42) [inline test/integration/scheduler/local-pv-neg-affinity_test.go]: does this need to be defined outside?

File: test/integration/scheduler/local-pv-neg-affinity_test.go [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+	for i := 0; i < len(nodeMarkers); i++ {
+		podName = "local-pod-" + strconv.Itoa(i+1)
+		pod := makePod(podName, config.ns, []string{"local-pvc"})
+		nodeMarkers[i].(func(*v1.Pod, string))(pod, "node-2")
+		// Create Pod
+		if _, err := config.client.CoreV1().Pods(config.ns).Create(pod); err != nil {
+			t.Fatalf("Failed to create Pod %q: %v", pod.Name, err)
+		}
+		// Give time to shceduler to attempt to schedule pod
+		if err := waitForPodToSchedule(config.client, pod); err == nil {
```
Comment (msau42) [inline test/integration/scheduler/local-pv-neg-affinity_test.go]: How long does this wait?

File: test/integration/scheduler/local-pv-neg-affinity_test.go [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+		selector := fields.Set{
+			"involvedObject.kind":      "Pod",
+			"involvedObject.name":      podName,
+			"involvedObject.namespace": config.ns,
+			"reason":                   "FailedScheduling",
+		}.AsSelector().String()
+		options := metav1.ListOptions{FieldSelector: selector}
+		events, err := config.client.CoreV1().Events(config.ns).List(options)
+		if err != nil {
+			t.Errorf("Failed to list events with error: %v", err)
```
Comment (msau42) [inline test/integration/scheduler/local-pv-neg-affinity_test.go]: this should probably be Fatalf

File: test/integration/scheduler/local-pv-neg-affinity_test.go [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+		}
+		if len(events.Items) == 0 {
+			t.Error("Failed, no expected events generated")
+		}
+		found := false
+		for _, e := range events.Items {
+			if strings.Contains(e.Message, "MatchNodeSelector") ||
```
Comment (msau42) [inline test/integration/scheduler/local-pv-neg-affinity_test.go]: Does it have to contain both of them, or just at least one?

File: test/integration/scheduler/local-pv-neg-affinity_test.go (hunk 1 above)
Comment (msau42) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: That's odd. Are the node names, node labels, etc exactly the same between the two?

File: test/integration/scheduler/volume_binding_test.go [hunk 8]
```diff
@@ -461,7 +461,7 @@ func makePod(name, ns string, pvcs []string) *v1.Pod {
 			Containers: []v1.Container{
 				{
 					Name:    "write-pod",
-					Image:   "k8s.gcr.io/busybox:1.24",
+					Image:   "gcr.io/google_containers/busybox:1.24",
```
Comment (msau42) [inline test/integration/scheduler/volume_binding_test.go]: Why did you need to change this?

Comment (sbezverk) [reply test/integration/scheduler/volume_binding_test.go]: See this PR: 
https://github.com/kubernetes/kubernetes/commit/e9dd8a68f60d2565069caed068f8fe4d493ea5ce#diff-c0800dd83834b3847acb7efd347d6d34

File: test/integration/scheduler/local-pv-neg-affinity_test.go (hunk 7 above)
Comment (sbezverk) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: At least one

File: test/integration/scheduler/local-pv-neg-affinity_test.go (hunk 6 above)
Comment (sbezverk) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: done

File: test/integration/scheduler/local-pv-neg-affinity_test.go (hunk 5 above)
Comment (sbezverk) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: 30 seconds defined in here: https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler/util.go#L359

File: test/integration/scheduler/local-pv-neg-affinity_test.go (hunk 7 above)
Comment (msau42) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: Should it be both?

File: test/integration/scheduler/volume_binding_test.go (hunk 8 above)
Comment (msau42) [reply test/integration/scheduler/volume_binding_test.go]: Odd that the revert did not roll this back too?

File: test/integration/scheduler/local-pv-neg-affinity_test.go (hunk 4 above)
Comment (sbezverk) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: not anymore I had some debug outside of for loop. fixed.

File: test/integration/scheduler/local-pv-neg-affinity_test.go (hunk 7 above)
Comment (sbezverk) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: Well, "and" works too, I just wanted to be more flexible. It was just strange that both mismatches generate the same scheduler errors. I thought that it would be fixed in future.

Comment (msau42) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: We're testing that the combination of both predicate failures causes a scheduling failure.

Comment (sbezverk) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: ok changed for "and"

**Comment by sbezverk** [comment]
@msau42 the issue is/was because of labels. You did not apply label kubernetes.io/hostname I needed for scheduling to work and I missed labels needed for your test to work. I suggest to modify the setup: 1 - unconditionally assign "kubernetes.io/hostname" label to node(s) on creation, 2 - add extra label parameter which will be added all nodes in addition to "standard labels". What do you think? The only negative thing I see in this approach is what is somebody will need to apply different labels to different nodes... or another possibility would be to add a new func to assign user defined labels as in your test case to apply to a specific node selected by name. Let me know what do you think.

**Comment by msau42** [comment]
I think changing my test to use "kubernetes.io/hostname" label is fine.

File: test/integration/scheduler/local-pv-neg-affinity_test.go [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+			t.Errorf("Failed as Pod %s was scheduled sucessfully but expected to fail", pod.Name)
+		}
+		selector := fields.Set{
+			"involvedObject.kind":      "Pod",
+			"involvedObject.name":      podName,
+			"involvedObject.namespace": config.ns,
+			"reason":                   "FailedScheduling",
+		}.AsSelector().String()
+		options := metav1.ListOptions{FieldSelector: selector}
+		events, err := config.client.CoreV1().Events(config.ns).List(options)
```
Comment (timothysc) [inline test/integration/scheduler/local-pv-neg-affinity_test.go]: Why are you relying on events (which is subject to change) here.  If the data is not encoded in pod status then that is a bug.

**Review by timothysc — CHANGES_REQUESTED** [review]
What is the primary motivation of moving this from an e2e to a integration test?  Did it routinely flake, and if so why?  

This test is beyond the typical simple integration tests, you are standing up an apiserver, a controller, and fake nodes, PVs and PVCs...  which sounds an awful lot like an e2e.

File: test/integration/scheduler/local-pv-neg-affinity_test.go (hunk 9 above)
Comment (sbezverk) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: @timothysc thanks for your review. I do not think POD status provides very granular failure. In this test case we test a very specific scenario and scheduler reaction on it. @msau42 Appreciate your input here.

**Comment by sbezverk** [comment]
@timothysc The reason this test was moved from e2e is with inability to build multi node environment. This test was always skipped during e2e runs as it needed to have at least 2 nodes. With integration framework we could easily have it without any extra compute nodes/vms.

**Comment by msau42** [comment]
The motivation behind moving this to an integration test is because you only need three components to run: the scheduler, api server and pv controller.  You do not actually need a real multi-node cluster with running kubelets to test this functionality.  The benefits are that this functionality gets tested more frequently, it uses up less resources, and it's faster.  It's a big benefit for local development.

File: test/integration/scheduler/local-pv-neg-affinity_test.go (hunk 9 above)
Comment (msau42) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: The pod status would just be "pending", which I think is too generic of an error that we want to be testing.

Comment (timothysc) [reply test/integration/scheduler/local-pv-neg-affinity_test.go]: There are errors should be encoded into the status.message - https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.9/#podstatus-v1-core  that indicate the failure.  I'm ok with the event for now, but please add a // TODO about reevaluating or removing the event requirement on the test.

**Comment by timothysc** [comment]
> The motivation behind moving this to an integration test is because you only need three components to run: the scheduler, api server and pv controller. You do not actually need a real multi-node cluster with running kubelets to test this functionality. The benefits are that this functionality gets tested more frequently, it uses up less resources, and it's faster. It's a big benefit for local development.

"Only..." 
I'm ok with adding an integration test, but you will absolutely want to verify this behavior on a live cluster and in the future I could also see it being a conformance test.  

I'm ok with adding an integration for faster verification, but not ok with removing the e2e test.

**Comment by msau42** [comment]
In the scheduler code [here](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/scheduler.go#L195), I see that it does record an event and also updates the pod status with the error message.  So the test can be changed to look at the Pod status message instead.

**Comment by sbezverk** [comment]
@timothysc Please check the latest. I restored e2e tests and reworked checking for POD status messages.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve 

Just a note - There will likely be a reduction over time on integration startup routines.
